src/OrchardCore.Modules/OrchardCore.Media/Shortcodes/ShareShortcodeProvider.cs

```diff
@@ -0,0 +1,174 @@
+            _httpContextAccessor = httpContextAccessor;
+            _options = options.Value;
+        }
+
+        #region https://stackoverflow.com/a/40361205/549306
+        private static readonly Uri SomeBaseUri = new Uri("http://canbeanything");
+
+        private static string GetFileNameFromUrl(string url)
+        {
+            Uri uri;
+            if (!Uri.TryCreate(url, UriKind.Absolute, out uri))
+            {
+                uri = new Uri(SomeBaseUri, url);
+            }
+
+            return Path.GetFileName(uri.LocalPath);
+        }
+
+        #endregion
+
+        public ValueTask<string> EvaluateAsync(string identifier, Arguments arguments, string content, Context context)
+        {
+            // we can't process what we don't understand, and we are discarding or filtering out anything we don't understand to keep things safe
+            if (!Shortcodes.Contains(identifier))
+            {
+                return Null;
+            }
+
+            // critical that at least the content or the href/url/source is set
+            if (String.IsNullOrWhiteSpace(content) && !IsSet(arguments, ShortcodeProperty.Url))
+            {
+                return ShareShortcode;
+            }
+
+            string url;
+            string href;
+            // sort out and infer as necessary the description (content) and resource location (url and href)
+            if (IsSet(arguments, ShortcodeProperty.Url))
+            {
+                url = UrlFixer(arguments.NamedOrDefault(ShortcodeProperty.Url.Name));
+                if (String.IsNullOrWhiteSpace(content))
+                {
+                    content = GetFileNameFromUrl(url).Split(FileExtensionSeperator)[0];
+                }
+            }
+            else
+            {
+                url = UrlFixer(content);
+                // because both the content and the href can't be empty, we presume the content was the href in this branch, and replace a shortened, friendlier name for the resulting content
+                content = GetFileNameFromUrl(content).Split(FileExtensionSeperator)[0];
+            }
+
+            // urlencode to fix space in path rendering
+            href = FormatAsHtmlElementAttribute(ShortcodeProperty.Url, Uri.EscapeUriString(url));
+            var saveAs = IsSet(arguments, ShortcodeProperty.Save) ? arguments.NamedOrDefault(ShortcodeProperty.Save.Name) : GetFileNameFromUrl(url);
+            saveAs = FormatAsHtmlElementAttribute(ShortcodeProperty.Save, saveAs);
+
+            // process contextually benign adornments
+            var classList = String.Empty;
+            var titleText = String.Empty;
+            if (arguments.Any())
+            {
+
+                classList = FormatArguement(arguments, ShortcodeProperty.Class);
+                titleText = FormatArguement(arguments, ShortcodeProperty.Tooltip);
+                //var queryStringParams = new Dictionary<string, string>();
```

Comment:
Are queryStringParams needed under any use case?  

I didn't think so, but left this from the copy-pasta-refacta for discussion.